spark/unitycatalog/src/test/java/io/sparkuctest/UCDeltaTableIntegrationBaseTest.java

```diff
@@ -0,0 +1,325 @@
+   * @param tableSchema The table schema (e.g., "id INT, name STRING")
+   * @param testCode The test function that receives the full table name
+   */
+  protected void withNewTable(String tableName, String tableSchema, TestCode testCode) throws Exception {
+    withNewTable(tableName, tableSchema, TableType.EXTERNAL, testCode);
+  }
+
+  /**
+   * Functional interface for test code that takes a temporary directory.
+   */
+  @FunctionalInterface
+  protected interface TempDirCode {
+    void run(File dir) throws Exception;
+  }
+
+  /**
+   * Functional interface for test code that takes a table name parameter.
+   */
+  @FunctionalInterface
+  protected interface TestCode {
+    void run(String tableName) throws Exception;
+  }
+
+  /**
+   * Interface defining the interface for executing SQL and verifying results.
+   *
+   * This abstraction allows tests to be independent of the execution engine,
+   * making it easy to test the same logic via different interfaces (Spark SQL, JDBC, etc.).
+   */
+  public interface SQLExecutor {
+    /**
+     * Execute a SQL statement and return the results.
+     *
+     * @param sql The SQL statement to execute
+     * @return The query results as a list of rows, where each row is a list of strings
+     */
+    List<List<String>> runSQL(String sql);
+
+    /**
+     * Read all data from a table and verify it matches the expected results.
+     *
+     * @param tableName The fully qualified table name
+     * @param expected The expected results as a list of rows
+     */
+    void checkTable(String tableName, List<List<String>> expected);
+
+    /**
+     * Execute a SQL query and verify the results match the expected output.
+     *
+     * @param sql The SQL query to execute
+     * @param expected The expected results as a list of rows
+     */
+    void checkWithSQL(String sql, List<List<String>> expected);
+  }
+
+  /**
+   * Default SQL executor implementation using SparkSession.
+   *
+   * This executor runs all SQL queries through Spark SQL and converts
+   * results to string lists for easy comparison.
+   */
+  public static class SparkSQLExecutor implements SQLExecutor {
+    private final SparkSession spark;
+
+    public SparkSQLExecutor(SparkSession spark) {
+      this.spark = spark;
+    }
+
+    @Override
+    public List<List<String>> runSQL(String sql) {
+      Dataset<Row> df = spark.sql(sql);
+      Row[] rows = (Row[]) df.collect();
+      return Arrays.stream(rows)
+          .map(row -> {
+            List<String> cells = new java.util.ArrayList<>();
+            for (int i = 0; i < row.length(); i++) {
+              cells.add(row.isNullAt(i) ? "null" : row.get(i).toString());
+            }
+            return cells;
+          })
+          .collect(Collectors.toList());
+    }
+
+    @Override
+    public void checkTable(String tableName, List<List<String>> expected) {
+      List<List<String>> actual = runSQL("SELECT * FROM " + tableName + " ORDER BY 1");
```

Comment:
for the `ORDER BY 1` ,  it will requires that the first column is always unique, otherwise it may produce ambigious listing, which may make this tests to be flaky, right ? 

For example,  if we have a `id int, name string` table. and the if we insert 3 rows:

```
(1, 'AAA')
(2,'BBB')
(1, 'CCC')
```

Then the results of the `runSQL` can be either `[(1, 'AAA'), (1, 'CCC'), (2, 'BBB')]` or `[(1,'CCC'), (1, 'AAA'), (2, 'BBB')]`,  which make it to be a flaky test, right ?

Reply:
lets fix this local problem when it comes. i think we need to make this check order independent. like a set comparison rather than a seq comparison.